src/snapred/backend/data/Indexer.py

```diff
@@ -0,0 +1,373 @@
+    This is intended to have responsibility over one state/resolution/workflow combination,
+    (e.g state abcd123efg, lite mode, calibration), and for that combination will keep tabs on
+    the index and the directory tree to determine appropriate version numbers.
+
+    All saving or loading of indices, records, or calculation parameter files should be
+    handled through the appropriate Indexer.
+
+    When saving, the Indexer will always save at the version attached to the object.
+    If this version is invalid, it will throw an error and not save.
+
+    The Indexer version list will only update when both a record and a corresponding inex entry
+    have been written.
+"""
+
+
+class IndexerType(StrEnum):
+    DEFAULT = ""
+    CALIBRATION = "Calibration"
+    NORMALIZATION = "Normalization"
+    REDUCTION = "Reduction"
+
+
+# the record type for each indexer type
+RECORD_TYPE = {
+    IndexerType.CALIBRATION: CalibrationRecord,
+    IndexerType.NORMALIZATION: NormalizationRecord,
+    IndexerType.REDUCTION: ReductionRecord,
+    IndexerType.DEFAULT: Record,
+}
+
+# the params type for each indexer type
+PARAMS_TYPE = {
+    IndexerType.CALIBRATION: Calibration,
+    IndexerType.NORMALIZATION: Normalization,
+    IndexerType.REDUCTION: CalculationParameters,
+    IndexerType.DEFAULT: CalculationParameters,
+}
+
+
+class Indexer:
+    rootDirectory: Path
+    index: Dict[int, IndexEntry]
+
+    indexerType: IndexerType
+
+    ## CONSTRUCTOR / DESTRUCTOR METHODS ##
+
+    @validate_call
+    def __init__(self, *, indexerType: IndexerType, directory: Path | str) -> None:
+        self.indexerType = indexerType
+        self.rootDirectory = Path(directory)
+        self.index = self.readIndex()
+        self.dirVersions = self.readDirectoryList()
+        self.reconcileIndexToFiles()
+
+    def __del__(self):
+        # define the index to automatically write itself whenever the program closes
+        if self.rootDirectory.exists():
+            self.reconcileIndexToFiles()
+            self.writeIndex()
+
+    def readDirectoryList(self):
+        # create the directory version list from the directory structure
+        versions = set()
+        for fname in self.rootDirectory.glob("v_*"):
+            if os.path.isdir(fname):
+                version = str(fname).split("_")[-1]
+                if version == VERSION_DEFAULT_NAME:
+                    version = VERSION_DEFAULT
+                version = int(version)
+                versions.add(version)
+        return versions
+
+    def reconcileIndexToFiles(self):
+        # if a directory has no entry in the index, create one for it
+        versions = self.readDirectoryList()
+        for version in versions:
+            if version not in self.index:
+                record = self.readRecord(version)
+                self.index[version] = self.indexEntryFromRecord(record)
+        # if a version exists in the index with no corresponding directory, delete it
```

Comment:
Maybe it should be marked as "Not Found"? What if someone accidentally deleted a folder?  This may that less apparent.

Reply:
If someone accidentally deleted a folder... then you no longer have that calibration, and there's no further purpose in retaining its index entry.